src/LogProviders/BackEnd/RoslynLogger.cs

```diff
@@ -0,0 +1,252 @@
+            _roslynTraceSource = null;
+
+            _build?.Finish(succeeded: true, time: DateTime.Now);
+            _build?.SetLogPath(listener.LogPath);
+
+            _dataSource.NotifyChange();
+
+            listener.Dispose();
+        }
+
+        private bool EnsureRoslynLogger()
+        {
+            if (_setLoggerCall == null)
+            {
+                try
+                {
+                    var assembly = GetAssembly("Microsoft.VisualStudio.LanguageServices");
+                    if (assembly == null)
+                    {
+                        return false;
+                    }
+
+                    var type = assembly.GetType("Microsoft.VisualStudio.LanguageServices.RoslynActivityLogger");
+                    if (type == null)
+                    {
+                        return false;
+                    }
+
+                    var setLoggerMethodInfo = type.GetMethod("SetLogger", BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public);
+                    if (setLoggerMethodInfo != null)
+                    {
+                        _setLoggerCall = (Action<TraceSource>)setLoggerMethodInfo.CreateDelegate(typeof(Action<TraceSource>));
+                    }
+
+                    var removeLoggerMethodInfo = type.GetMethod("RemoveLogger", BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public);
+                    if (removeLoggerMethodInfo != null)
+                    {
+                        _removeLoggerCall = (Action<TraceSource>)removeLoggerMethodInfo.CreateDelegate(typeof(Action<TraceSource>));
+                    }
+                }
+                catch
+                {
+                    return false;
+                }
+            }
+
+            return _setLoggerCall != null;
+        }
+
+        private static Assembly GetAssembly(string assemblyName) =>
+            (from assembly in AppDomain.CurrentDomain.GetAssemblies()
+                where assembly.FullName.IndexOf(assemblyName, StringComparison.OrdinalIgnoreCase) >= 0
+                let current = assembly.GetName()
+                where string.Equals(current.Name, assemblyName, StringComparison.OrdinalIgnoreCase)
+                select assembly).FirstOrDefault();
+
+        private class RoslynTraceListener : TraceListener
+        {
+            private const int LogEventId = 0;
+            private const int StartEventId = 1;
+            private const int EndEventId = 2;
+
+            private readonly IImmutableSet<string> _set;
+            private readonly StreamWriter _writer;
+
+            public RoslynTraceListener(IImmutableSet<string> roslynEvents)
+            {
+                LogPath = Path.Combine(Path.GetTempPath(), $"RoslynLog-{Guid.NewGuid()}.txt");
+
+                _set = roslynEvents;
+
+                // if there is an issue, this will crash VS. we can swallow, but then, user won't know
```

Comment:
>this will crash VS [](start = 41, length = 18)

Since this is going in a server can we make sure this won't crash VS?

Reply:
Done